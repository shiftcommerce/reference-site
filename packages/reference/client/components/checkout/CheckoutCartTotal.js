// Libraries
import { Component } from 'react'
import Sticky from 'react-stickyfill'
import Link from 'next/link'

// Lib
import { calculateCartSummary } from '../../lib/calculateCartSummary'
import { fixedPrice } from '../../lib/fixedPrice'

// Objects
import Button from '../../objects/Button'

class CheckoutCartTotal extends Component {
  validOrder (checkout, order) {
    const checkoutSteps = [ 'shippingAddress', 'shippingMethod', 'paymentMethod' ]
    const cardErrors = order.card_errors
    const allStepsCompleted = checkoutSteps.every((step) => (checkout[step].completed))

    if (!cardErrors && allStepsCompleted) {
      return allStepsCompleted
    } else { return false }
  }

  renderButtons () {
    const { checkout, onClick, order } = this.props
    const isValidOrder = this.validOrder(checkout, order)

    if (checkout.currentStep === 3) {
      return <>
        <div className='c-cart-summary__buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              status='primary'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              type='button'
            />
          </Link>
          <Button
            aria-label='Review Order'
            label='Review Your Order'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            onClick={onClick}
          />
        </div >
      </>
    } else if (checkout.currentStep === 4) {
      return <>
        <div className='c-cart-summary-buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              status='primary'
              type='button'
            />
          </Link>
          <Button
            aria-label='Place Order'
            label='Place Order'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            id='place_order'
            status={(isValidOrder ? 'primary' : 'disabled')}
            disabled={!isValidOrder}
            onClick={onClick}
          />
        </div >
      </>
    } else {
      return <>
        <div className='c-cart-summary-buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              status='primary'
              type='button'
            />
          </Link>
          <Button
            aria-label='Continue to Payment'
            label='Continue to Payment'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            onClick={onClick}
          />
        </div >
      </>
    }
  }

  render () {
    const { cart, checkout, order } = this.props
    const totals = calculateCartSummary(cart, checkout)

    let shippingText = ''
    if (!checkout.shippingAddress.completed) {
      shippingText = 'Enter address'
    } else if (!checkout.shippingMethod.retail_price_inc_tax) {
      shippingText = 'Select shipping method'
    } else {
      shippingText = `Â£${fixedPrice(totals.shipping)}`
    }

    return (
      <>
      <Sticky>
        <div aria-label='Cart total summary' className='u-sticky  c-cart-summary'>
          <div className=''>
            <dl aria-label='Subtotal'>
              <dt> Total Price: </dt>
              <dd> &pound;{ fixedPrice(totals.subTotal) } </dd>
            </dl>
            <dl aria-label='Shipping cost'>
              <dt> Shipping costs: </dt>
              <dd> { shippingText } </dd>
            </dl>
            <dl aria-label='Total' className='u-bold'>
              <dt> TOTAL: </dt>
              <dd> &pound;{ fixedPrice(totals.total) } </dd>
            </dl>
            <dl>
              <dt className='c-cart-summary__VAT'>* Including VAT</dt>
            </dl>
            { order.paymentError &&
              <div className='c-checkout-cart-total__payment-error'>{ order.paymentError }</div>
            }
          </div>
        </div>
      </Sticky>
      { this.renderButtons() }
      </>
    )
  }
}

export default CheckoutCartTotal
