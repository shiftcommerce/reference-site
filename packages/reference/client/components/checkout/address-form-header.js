import Router from 'next/router'
import { Button } from 'shift-react-components'

export default function AddressFormHeader ({ collapsed, title }) {
  return (
    <div className='o-form__header c-address-form__header'>
      <div className='o-form__header-title c-address-form__header-title'>
        <h2>{ title }</h2>
      </div>
      { collapsed &&
          <Button
            label='Edit'
            status='secondary'
            className='o-button-edit'
            onClick={() => Router.push('/checkout/shipping-address')}
          />
      }
    </div>
  )
}
