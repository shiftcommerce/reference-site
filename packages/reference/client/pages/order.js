//  Libraries
import { connect } from 'react-redux'

// Pages
import { algoliaReduxWrapper } from '@shiftcommerce/shift-next/src/lib/algolia-redux-wrapper'
import OrderPage from '@shiftcommerce/shift-next/src/pages/order'

const mapStateToProps = (state) => {
  const { order } = state
  return { order }
}

export default algoliaReduxWrapper(connect(mapStateToProps)(OrderPage), OrderPage)
