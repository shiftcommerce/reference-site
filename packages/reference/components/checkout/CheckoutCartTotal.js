// Libraries
import { Component } from 'react'
import { calculateCartSummary } from '../../lib/calculateCartSummary'

class CheckoutCartTotal extends Component {
  render () {
    const {subTotal, total, tax} = calculateCartSummary(this.props.cart)

    return (
      <div aria-label='Cart total summary' className='o-checkout-cart-total'>
        <div className='o-checkout-cart-total__wrapper'>
          <dl aria-label='Subtotal'>
            <dt> Subtotal: </dt>
            <dd> &pound;{subTotal} </dd>
          </dl>
          <dl aria-label='VAT'>
            <dt> VAT: </dt>
            <dd> &pound;{tax} </dd>
          </dl>
          <dl aria-label='Shipping cost'>
            <dt> Shipping: </dt>
            <dd> Enter address </dd>
          </dl>
          <dl aria-label='Total' className='o-checkout-cart-total__total'>
            <dt> You Pay: </dt>
            <dd> <b>&pound;{total}</b> </dd>
          </dl>
        </div>
      </div>
    )
  }
}

export default CheckoutCartTotal
