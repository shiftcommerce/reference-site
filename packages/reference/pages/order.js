//  React
import { Component } from 'react'

// Redux
import withRedux from 'next-redux-wrapper'

// Components
import Layout from '../components/Layout'

// Utils
import { configureStore } from '../utils/configureStore'

class OrderPage extends Component {
  render () {
    const order = this.props.order

    return (
      <Layout>
        Order has been created: #{order.id}
      </Layout>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    order: state.order || {}
  }
}

export default withRedux(configureStore, mapStateToProps)(OrderPage)
