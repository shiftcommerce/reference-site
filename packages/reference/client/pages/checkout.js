// Libraries
import { Component } from 'react'
import { connect } from 'react-redux'
import Router from 'next/router'

// Libs
import InputFieldValidator from '../lib/input-field-validator'

// Actions
import { readCheckoutFromLocalStorage,
  toggleCollapsed,
  setShippingMethod,
  changeBillingAddress,
  setValidationMessage,
  changePaymentMethod,
  inputChange,
  inputComplete,
  showField } from '../actions/checkout-actions'
import { readCart, initializeCart, updateQuantity } from '../actions/cart-actions'
import {
  createOrder,
  requestCardToken,
  setCardToken,
  setPaymentError,
  setCardErrors
} from '../actions/order-actions'

// Objects
import Logo from '../objects/logo'

// Components
import CustomHead from '../components/custom-head'
import CheckoutSteps from '../components/checkout/checkout-steps'
import AddressForm from '../components/checkout/address-form'
import CheckoutCart from '../components/checkout/checkout-cart'
import CheckoutCartTotal from '../components/checkout/checkout-cart-total'
import ShippingMethods from '../components/checkout/shipping-methods'
import PaymentMethod from '../components/checkout/payment-method'
import PromoInput from '../components/promo-input'

export class CheckoutPage extends Component {
  static async getInitialProps ({ reduxStore }) {
    return {}
  }
  constructor () {
    super()

    this.onToggleCollapsed = this.onToggleCollapsed.bind(this)
    this.nextSection = this.nextSection.bind(this)
    this.setShippingMethod = this.setShippingMethod.bind(this)
    this.onInputChange = this.onInputChange.bind(this)
    this.onInputBlur = this.onInputBlur.bind(this)
    this.onShowField = this.onShowField.bind(this)
    this.changeBillingAddress = this.changeBillingAddress.bind(this)
    this.convertToOrder = this.convertToOrder.bind(this)
    this.onPaymentMethodChanged = this.onPaymentMethodChanged.bind(this)
    this.onCardTokenReceived = this.onCardTokenReceived.bind(this)
    this.setCardErrors = this.setCardErrors.bind(this)
    this.updateQuantity = this.updateQuantity.bind(this)
    this.deleteItem = this.deleteItem.bind(this)
  }

  componentDidMount () {
    this.props.dispatch(readCheckoutFromLocalStorage())
    this.props.dispatch(readCart()).then((cart) => {
      if (cart.totalQuantity === 0) Router.push('/cart')
    })
  }

  componentWillReceiveProps (props) {
    const prevOrder = this.props.order
    const newOrder = props.order
    // Only when new order got created,
    // Redirect to order confirmation page
    if (newOrder && newOrder.id && newOrder.id !== prevOrder.id) {
      Router.push('/order')
      props.dispatch(initializeCart())
    }
  }

  updateQuantity (e) {
    const lineItem = {
      sku: e.target.dataset.variant,
      quantity: parseInt(e.target.value, 10)
    }
    this.props.dispatch(updateQuantity(lineItem))
  }

  deleteItem (e) {
    e.preventDefault()
    const lineItem = {
      sku: e.target.dataset.variant,
      quantity: 0
    }
    this.props.dispatch(updateQuantity(lineItem))
    if (this.props.cart.lineItems.length === 0) {
      Router.push('/cart')
    }
  }

  nextSection (eventType) {
    const { checkout } = this.props

    let componentName

    if (checkout.currentStep === 1) {
      componentName = 'shippingAddress'
    } else if (checkout.currentStep === 2) {
      componentName = 'shippingMethod'
    } else if (checkout.currentStep === 3) {
      componentName = 'paymentMethod'
    }

    window.scrollTo(0, 0)
    this.props.dispatch(toggleCollapsed(eventType, componentName))
  }

  onToggleCollapsed (eventType, componentName) {
    window.scrollTo(0, 0)
    this.props.dispatch(toggleCollapsed(eventType, componentName))
  }

  setShippingMethod (shippingMethod) {
    this.props.dispatch(setShippingMethod(shippingMethod))
  }

  onInputChange (event, formName, fieldName, fieldValue) {
    this.props.dispatch(inputChange(formName, fieldName, fieldValue))
  }

  validateInput (formName, fieldName, fieldValue, rules) {
    let validationMessage = new InputFieldValidator(fieldName, fieldValue, rules).validate()
    this.props.dispatch(setValidationMessage(formName, fieldName, validationMessage))
  }

  onInputBlur (event, formName, fieldName, fieldValue, rules) {
    this.validateInput(formName, fieldName, fieldValue, rules)
    this.props.dispatch(inputComplete())
  }

  convertToOrder (event) {
    if (this.props.checkout.paymentMethod.selectedMethod === 'card') {
      this.props.dispatch(requestCardToken(true))
    } else {
      this.props.dispatch(createOrder(this.props.cart, this.props.checkout, this.props.order))
    }
  }

  onShowField (formName, fieldName) {
    this.props.dispatch(showField(formName, fieldName))
  }

  onPaymentMethodChanged (paymentMethod) {
    this.props.dispatch(changePaymentMethod(paymentMethod))
  }

  onCardTokenReceived ({ error, token }) {
    if (error) {
      this.props.dispatch(setPaymentError(error.message))
    } else {
      this.props.dispatch(setCardToken(token))
    }
    this.props.dispatch(requestCardToken(false))
  }

  setCardErrors (error) {
    this.props.dispatch(setCardErrors(error))
  }

  changeBillingAddress (event, formName, fieldName) {
    this.props.dispatch(changeBillingAddress(formName, fieldName, event.target.checked))
  }

  isPaymentValid () {
    let billingAddressErrors = this.props.checkout.billingAddress.errors
    const isCardValid = !this.props.order.cardErrors
    const isBillingAddressValid = Object.keys(billingAddressErrors).every((key) => billingAddressErrors[key] === '') === true
    return isBillingAddressValid && isCardValid
  }

  render () {
    const { checkout, cart } = this.props
    const hasLineItems = cart.totalQuantity > 0
    if (checkout.loading) {
      return (
        <div>
          <CustomHead />
          loading...
        </div>
      )
    } else if (checkout.error) {
      return (
        <p>{ checkout.error }</p>
      )
    } else {
      return (
        <>
          <CustomHead />
          {hasLineItems &&
            <>
              <div className='o-header'>
                <div className='c-step-indicators__logo'>
                  <Logo className='u-text-color--primary' />
                </div>
                <div>
                  <CheckoutSteps {...this.props} />
                </div>
              </div>
              <div className='o-grid-container'>
                <div className='o-col-1-13 o-col-1-9-l'>
                  <AddressForm {...this.props}
                    title='Shipping Address'
                    formName='shippingAddress'
                    addressType='shipping'
                    onChange={this.onInputChange}
                    onBlur={this.onInputBlur}
                    onShowField={this.onShowField}
                    onToggleCollapsed={this.onToggleCollapsed}
                  />
                  <ShippingMethods
                    {...this.props}
                    formName='shippingMethod'
                    setShippingMethod={this.setShippingMethod}
                    onToggleCollapsed={this.onToggleCollapsed}
                  />
                  <PaymentMethod
                    formName='paymentMethod'
                    {...this.props}
                    changeBillingAddress={this.changeBillingAddress}
                    onChange={this.onInputChange}
                    onBlur={this.onInputBlur}
                    onPaymentMethodChanged={this.onPaymentMethodChanged}
                    onToggleCollapsed={this.onToggleCollapsed}
                    onCardTokenReceived={this.onCardTokenReceived}
                    setCardErrors={this.setCardErrors}
                  />
                </div>
                <PromoInput />
                <div className='o-col-1-13 o-col-9-13-l c-cart-table'>
                  <CheckoutCart title='Your Cart' {...this.props} updateQuantity={this.updateQuantity} deleteItem={this.deleteItem} />
                  <CheckoutCartTotal {...this.props} convertToOrder={this.convertToOrder} onClick={() => { this.nextSection('complete') }} />
                </div>
              </div>
            </>
          }
        </>
      )
    }
  }
}

const mapStateToProps = (state) => {
  return {
    checkout: state.checkout || {},
    cart: state.cart || {},
    order: state.order || {},
    ip: state.ip
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatch: dispatch
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckoutPage)
