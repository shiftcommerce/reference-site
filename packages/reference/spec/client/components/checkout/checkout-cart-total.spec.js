// Components
import CheckoutCartTotal from '../../../../client/components/checkout/checkout-cart-total'

// Fixtures
import checkout from '../../../fixtures/checkout'
import order from '../../../fixtures/order'

test("renders the cart's total, subtotal and shipping cost", () => {
  // arrange
  const cart = {
    sub_total: 15,
    total: 20,
    discount_summaries: [],
    shipping_method: {
      total: 5
    }
  }

  const checkout = {
    shippingAddress: {}
  }

  const order = {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} order={order} title='Checkout Cart Summary' />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Total Price:  £15.00')
  expect(wrapper).toIncludeText('Shipping costs:  £5.00')
  expect(wrapper).toIncludeText('TOTAL:  £20.00')
})

test('renders promotions', () => {
  // arrange
  const cart = {
    sub_total: 15,
    total: 20,
    discount_summaries: [{
      id: 1,
      name: 'Christmas Special',
      total: 1.5
    }],
    shipping_method: {
      total: 5
    }
  }

  const checkout = {
    shippingAddress: {}
  }

  const order = {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} order={order} title='Checkout Cart Summary' />
  )

  // assert
  expect(wrapper).toIncludeText('Christmas Special:- £1.50')
})

test('renders shipping promotions', () => {
  // arrange
  const cart = {
    sub_total: 15,
    total: 20,
    shipping_method: {
      total: 5
    },
    discount_summaries: [],
    shipping_discount_name: 'Free shipping',
    shipping_total_discount: -5
  }

  const checkout = {
    shippingAddress: {}
  }

  const order = {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} order={order} title='Checkout Cart Summary' />
  )

  // assert
  expect(wrapper).toIncludeText('Free shipping:- £5.00')
})

test('renders payment errors', () => {
  // arrange
  const cart = {
    sub_total: 15,
    total: 20,
    discount_summaries: [],
    shipping_method: {
      total: 5
    }
  }

  const checkout = {
    shippingAddress: {}
  }

  const order = {
    paymentError: 'Payment authentication failed'
  }

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} order={order} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Payment authentication failed')
})

test('displays a correct disabled button when shipping address is not completed yet', () => {
  const checkout = {
    currentStep: 1,
    shippingAddress: {
      errors: []
    }
  }

  const cart = {
    sub_total: 15,
    total: 20,
    discount_summaries: []
  }

  const wrapper = mount(
    <CheckoutCartTotal checkout={checkout} order={{}} cart={cart} dispatch={jest.fn()} />
  )

  expect(wrapper.find('button.c-cart-summary__buttons--proceed')).toHaveText('Choose Shipping Method')
  expect(wrapper.find('button.c-cart-summary__buttons--proceed')).toHaveProp('disabled', true)
})

test('displays a correct enabled button when shipping address is completed', () => {
  const checkout = {
    currentStep: 1,
    shippingAddress: {
      first_name: 'John',
      last_name: 'Doe',
      primary_phone: '07510123123',
      email: 'john@example.com',
      country_code: 'UK',
      line_1: '1 Line',
      zipcode: 'Ls27EY',
      city: 'Leeds',
      errors: {}
    }
  }

  const cart = {
    sub_total: 15,
    total: 20,
    discount_summaries: []
  }

  const wrapper = mount(
    <CheckoutCartTotal checkout={checkout} order={{}} cart={cart} dispatch={jest.fn()} />
  )

  expect(wrapper.find('button.c-cart-summary__buttons--proceed')).toHaveText('Choose Shipping Method')
  expect(wrapper.find('button.c-cart-summary__buttons--proceed')).toHaveProp('disabled', false)
})

describe('Place Order button', () => {
  test('should be disabled if there is any error in billing address', () => {
    // Arrange
    const dispatch = jest.fn().mockImplementation((test) => Promise.resolve())
    const newCheckout = Object.assign({}, checkout, {
      paymentMethod: {
        ...checkout.paymentMethod,
        collapsed: false
      },
      billingAddress: {
        ...checkout.billingAddress,
        errors: {
          first_name: 'please fill the details'
        }
      },
      currentStep: 4
    })

    const cart = {
      sub_total: 15,
      total: 20,
      discount_summaries: []
    }

    // Act
    const wrapper = mount(
      <CheckoutCartTotal checkout={newCheckout} order={order} cart={cart} dispatch={dispatch} />
    )

    // Assert
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('button#place_order')).toBeDisabled()
  })

  test('should be disabled if there is any error in card details', () => {
    // Arrange
    const dispatch = jest.fn().mockImplementation((test) => Promise.resolve())
    const newCheckout = Object.assign({}, checkout, {
      paymentMethod: {
        ...checkout.paymentMethod,
        collapsed: false
      },
      currentStep: 4
    })

    const newOrder = {
      ...order,
      cardErrors: true
    }

    const cart = {
      sub_total: 15,
      total: 20,
      discount_summaries: []
    }

    // Act
    const wrapper = mount(
      <CheckoutCartTotal checkout={newCheckout} order={newOrder} cart={cart} dispatch={dispatch} />
    )

    // Assert
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('button#place_order')).toBeDisabled()
  })
})
