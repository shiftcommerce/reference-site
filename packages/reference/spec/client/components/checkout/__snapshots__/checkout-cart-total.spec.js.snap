// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Place Order button should be disabled if there is any error in billing address 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "discount_summaries": Array [],
      "sub_total": 15,
      "total": 20,
    }
  }
  checkout={
    Object {
      "billingAddress": Object {
        "city": "Leeds",
        "collapsed": true,
        "completed": true,
        "country_code": "GBP",
        "errors": Object {
          "first_name": "please fill the details",
        },
        "first_name": "Prasanna Gaddam",
        "label": "BILLING_LABEL",
        "last_name": "Prasanna Gaddam",
        "line_1": "Apartment 19, Waterloo Court, 17 Hunslet Road",
        "line_2": "Hunslet",
        "primary_phone": "07400000000",
        "state": "Yorkshire",
        "zipcode": "LS10 1QN",
      },
      "currency": "GBP",
      "currentStep": 4,
      "customer": Object {
        "email": "customer@example.com",
        "external_id": "123",
        "name": "John Doe",
      },
      "error": false,
      "lineItems": Array [
        Object {
          "currency": "GBP",
          "paid_price_inc_tax": "10",
          "quantity": 3,
          "retail_price_inc_tax": "10",
          "sku": "SK840763631-2",
          "tax_amount": 0,
          "tax_percentage": 0,
          "title": "Textured Long T Shirt",
        },
      ],
      "loading": false,
      "paid_price_inc_tax": "0105.45",
      "paymentMethod": Object {
        "collapsed": false,
        "completed": false,
      },
      "request_uuid": 6765757657657,
      "retail_price_inc_tax": "0105.45",
      "reviewOrder": Object {
        "collapsed": true,
        "completed": false,
      },
      "shippingAddress": Object {
        "city": "Leeds",
        "collapsed": true,
        "completed": true,
        "country_code": "GBP",
        "first_name": "Prasanna Gaddam",
        "label": "BILLING_LABEL",
        "last_name": "Prasanna Gaddam",
        "line_1": "Apartment 19, Waterloo Court, 17 Hunslet Road",
        "line_2": "Hunslet",
        "primary_phone": "07400000000",
        "state": "Yorkshire",
        "zipcode": "LS10 1QN",
      },
      "shippingMethod": Object {
        "collapsed": false,
        "completed": true,
        "id": "2",
        "name": "UPS Next Day",
        "paid_price_inc_tax": 5.45,
        "reference": "ups_next_day",
        "retail_price_inc_tax": 5.45,
        "sku": "ups_next_day",
        "tax_amount": 2,
        "tax_percentage": 20,
      },
      "tax_amount": 0,
    }
  }
  dispatch={[MockFunction]}
  order={
    Object {
      "cardErrors": false,
      "cardTokenRequested": false,
    }
  }
>
  <Sticker>
    <div
      aria-label="Cart total summary"
      className="u-sticky c-cart-summary"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Total Price: 
        </dt>
        <dd>
           £
          15.00
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping costs: 
        </dt>
        <dd>
           
          Loading shipping cost...
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="u-bold"
      >
        <dt>
           TOTAL: 
        </dt>
        <dd>
           £
          20.00
           
        </dd>
      </dl>
      <dl>
        <dt
          className="c-cart-summary__VAT"
        >
          * Including VAT
        </dt>
      </dl>
    </div>
  </Sticker>
  <div
    className="c-cart-summary-buttons"
  >
    <Link
      href="/"
    >
      <Button
        aria-label="Continue Shopping"
        className="c-cart-summary__buttons--continue o-button--sml"
        label="Continue shopping"
        onClick={[Function]}
        status="primary"
        type="button"
      >
        <div
          className="o-button__container"
        >
          <button
            aria-label="Continue Shopping"
            className="c-cart-summary__buttons--continue o-button--sml o-button o-button--primary"
            disabled={false}
            onClick={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            Continue shopping
          </button>
        </div>
      </Button>
    </Link>
    <Button
      aria-label="Place Order"
      className="c-cart-summary__buttons--proceed o-button--sml"
      disabled={true}
      id="place_order"
      label="Place Order"
      status="disabled"
      type="button"
    >
      <div
        className="o-button__container"
      >
        <button
          aria-label="Place Order"
          className="c-cart-summary__buttons--proceed o-button--sml o-button o-button--disabled o-button--disabled"
          disabled={true}
          id="place_order"
          role="button"
          tabIndex="0"
          type="button"
        >
          Place Order
        </button>
      </div>
    </Button>
  </div>
</CheckoutCartTotal>
`;

exports[`Place Order button should be disabled if there is any error in card details 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "discount_summaries": Array [],
      "sub_total": 15,
      "total": 20,
    }
  }
  checkout={
    Object {
      "billingAddress": Object {
        "city": "Leeds",
        "collapsed": true,
        "completed": true,
        "country_code": "GBP",
        "errors": Object {},
        "first_name": "Prasanna Gaddam",
        "label": "BILLING_LABEL",
        "last_name": "Prasanna Gaddam",
        "line_1": "Apartment 19, Waterloo Court, 17 Hunslet Road",
        "line_2": "Hunslet",
        "primary_phone": "07400000000",
        "state": "Yorkshire",
        "zipcode": "LS10 1QN",
      },
      "currency": "GBP",
      "currentStep": 4,
      "customer": Object {
        "email": "customer@example.com",
        "external_id": "123",
        "name": "John Doe",
      },
      "error": false,
      "lineItems": Array [
        Object {
          "currency": "GBP",
          "paid_price_inc_tax": "10",
          "quantity": 3,
          "retail_price_inc_tax": "10",
          "sku": "SK840763631-2",
          "tax_amount": 0,
          "tax_percentage": 0,
          "title": "Textured Long T Shirt",
        },
      ],
      "loading": false,
      "paid_price_inc_tax": "0105.45",
      "paymentMethod": Object {
        "collapsed": false,
        "completed": false,
      },
      "request_uuid": 6765757657657,
      "retail_price_inc_tax": "0105.45",
      "reviewOrder": Object {
        "collapsed": true,
        "completed": false,
      },
      "shippingAddress": Object {
        "city": "Leeds",
        "collapsed": true,
        "completed": true,
        "country_code": "GBP",
        "first_name": "Prasanna Gaddam",
        "label": "BILLING_LABEL",
        "last_name": "Prasanna Gaddam",
        "line_1": "Apartment 19, Waterloo Court, 17 Hunslet Road",
        "line_2": "Hunslet",
        "primary_phone": "07400000000",
        "state": "Yorkshire",
        "zipcode": "LS10 1QN",
      },
      "shippingMethod": Object {
        "collapsed": false,
        "completed": true,
        "id": "2",
        "name": "UPS Next Day",
        "paid_price_inc_tax": 5.45,
        "reference": "ups_next_day",
        "retail_price_inc_tax": 5.45,
        "sku": "ups_next_day",
        "tax_amount": 2,
        "tax_percentage": 20,
      },
      "tax_amount": 0,
    }
  }
  dispatch={[MockFunction]}
  order={
    Object {
      "cardErrors": true,
      "cardTokenRequested": false,
    }
  }
>
  <Sticker>
    <div
      aria-label="Cart total summary"
      className="u-sticky c-cart-summary"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Total Price: 
        </dt>
        <dd>
           £
          15.00
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping costs: 
        </dt>
        <dd>
           
          Loading shipping cost...
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="u-bold"
      >
        <dt>
           TOTAL: 
        </dt>
        <dd>
           £
          20.00
           
        </dd>
      </dl>
      <dl>
        <dt
          className="c-cart-summary__VAT"
        >
          * Including VAT
        </dt>
      </dl>
    </div>
  </Sticker>
  <div
    className="c-cart-summary-buttons"
  >
    <Link
      href="/"
    >
      <Button
        aria-label="Continue Shopping"
        className="c-cart-summary__buttons--continue o-button--sml"
        label="Continue shopping"
        onClick={[Function]}
        status="primary"
        type="button"
      >
        <div
          className="o-button__container"
        >
          <button
            aria-label="Continue Shopping"
            className="c-cart-summary__buttons--continue o-button--sml o-button o-button--primary"
            disabled={false}
            onClick={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            Continue shopping
          </button>
        </div>
      </Button>
    </Link>
    <Button
      aria-label="Place Order"
      className="c-cart-summary__buttons--proceed o-button--sml"
      disabled={true}
      id="place_order"
      label="Place Order"
      status="disabled"
      type="button"
    >
      <div
        className="o-button__container"
      >
        <button
          aria-label="Place Order"
          className="c-cart-summary__buttons--proceed o-button--sml o-button o-button--disabled o-button--disabled"
          disabled={true}
          id="place_order"
          role="button"
          tabIndex="0"
          type="button"
        >
          Place Order
        </button>
      </div>
    </Button>
  </div>
</CheckoutCartTotal>
`;

exports[`renders payment errors 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "discount_summaries": Array [],
      "shipping_method": Object {
        "total": 5,
      },
      "sub_total": 15,
      "total": 20,
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {},
    }
  }
  order={
    Object {
      "paymentError": "Payment authentication failed",
    }
  }
>
  <Sticker>
    <div
      aria-label="Cart total summary"
      className="u-sticky c-cart-summary"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Total Price: 
        </dt>
        <dd>
           £
          15.00
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping costs: 
        </dt>
        <dd>
           
          £5.00
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="u-bold"
      >
        <dt>
           TOTAL: 
        </dt>
        <dd>
           £
          20.00
           
        </dd>
      </dl>
      <dl>
        <dt
          className="c-cart-summary__VAT"
        >
          * Including VAT
        </dt>
      </dl>
      <div
        className="c-checkout-cart-total__payment-error"
      >
        Payment authentication failed
      </div>
    </div>
  </Sticker>
  <div
    className="c-cart-summary-buttons"
  >
    <Link
      href="/"
    >
      <Button
        aria-label="Continue Shopping"
        className="c-cart-summary__buttons--continue o-button--sml"
        label="Continue shopping"
        onClick={[Function]}
        status="primary"
        type="button"
      >
        <div
          className="o-button__container"
        >
          <button
            aria-label="Continue Shopping"
            className="c-cart-summary__buttons--continue o-button--sml o-button o-button--primary"
            disabled={false}
            onClick={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            Continue shopping
          </button>
        </div>
      </Button>
    </Link>
    <Button
      className="c-cart-summary__buttons--proceed o-button--sml"
      type="button"
    >
      <div
        className="o-button__container"
      >
        <button
          className="c-cart-summary__buttons--proceed o-button--sml o-button"
          disabled={false}
          role="button"
          tabIndex="0"
          type="button"
        />
      </div>
    </Button>
  </div>
</CheckoutCartTotal>
`;

exports[`renders the cart's total, subtotal and shipping cost 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "discount_summaries": Array [],
      "shipping_method": Object {
        "total": 5,
      },
      "sub_total": 15,
      "total": 20,
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {},
    }
  }
  order={Object {}}
  title="Checkout Cart Summary"
>
  <Sticker>
    <div
      aria-label="Cart total summary"
      className="u-sticky c-cart-summary"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Total Price: 
        </dt>
        <dd>
           £
          15.00
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping costs: 
        </dt>
        <dd>
           
          £5.00
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="u-bold"
      >
        <dt>
           TOTAL: 
        </dt>
        <dd>
           £
          20.00
           
        </dd>
      </dl>
      <dl>
        <dt
          className="c-cart-summary__VAT"
        >
          * Including VAT
        </dt>
      </dl>
    </div>
  </Sticker>
  <div
    className="c-cart-summary-buttons"
  >
    <Link
      href="/"
    >
      <Button
        aria-label="Continue Shopping"
        className="c-cart-summary__buttons--continue o-button--sml"
        label="Continue shopping"
        onClick={[Function]}
        status="primary"
        type="button"
      >
        <div
          className="o-button__container"
        >
          <button
            aria-label="Continue Shopping"
            className="c-cart-summary__buttons--continue o-button--sml o-button o-button--primary"
            disabled={false}
            onClick={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            Continue shopping
          </button>
        </div>
      </Button>
    </Link>
    <Button
      className="c-cart-summary__buttons--proceed o-button--sml"
      type="button"
    >
      <div
        className="o-button__container"
      >
        <button
          className="c-cart-summary__buttons--proceed o-button--sml o-button"
          disabled={false}
          role="button"
          tabIndex="0"
          type="button"
        />
      </div>
    </Button>
  </div>
</CheckoutCartTotal>
`;
